File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/vector/VectorSimilarityFunction.java
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.neo4j.kernel.api.impl.schema.vector;

import java.util.EnumSet;
import java.util.Locale;
import org.neo4j.kernel.api.vector.VectorCandidate;

public enum VectorSimilarityFunction {
    // TODO VECTOR: perhaps some unrolling and/or vector api (when available) could be used here
    //              perhaps investigate some more accurate normalisation techniques

    EUCLIDEAN {
        @Override
        public float[] maybeToValidVector(VectorCandidate candidate) {
            final int dimensions;
            if (candidate == null || (dimensions = candidate.dimensions()) == 0) {
                return null;
            }

            final var vector = new float[dimensions];
            for (int i = 0; i < dimensions; i++) {
                final var element = candidate.floatElement(i);
                if (!Float.isFinite(element)) {
                    return null;
                }
                vector[i] = element;
            }
            return vector;
        }
    },

    COSINE {
        @Override
        public float[] maybeToValidVector(VectorCandidate candidate) {
            final int dimensions;
            if (candidate == null || (dimensions = candidate.dimensions()) == 0) {
                return null;
            }

            float square = 0.f;
            final var vector = new float[dimensions];
            for (int i = 0; i < dimensions; i++) {
                final var element = candidate.floatElement(i);
                if (!Float.isFinite(element)) {
                    return null;
                }
                square += element * element;
                vector[i] = element;
            }

            if (square <= 0.f || !Float.isFinite(square)) {
                return null;
            }

            return vector;
        }
    };

    public static final EnumSet<VectorSimilarityFunction> SUPPORTED = EnumSet.allOf(VectorSimilarityFunction.class);

    public static VectorSimilarityFunction fromName(String name) {
        try {
            return valueOf(name.toUpperCase(Locale.ROOT));
        } catch (IllegalArgumentException e) {
            final var exception = new IllegalArgumentException(
                    "'%s' is an unsupported vector similarity function. Supported: %s".formatted(name, SUPPORTED));
            exception.addSuppressed(e);
            throw exception;
        }
    }

    public final float[] maybeToValidVector(Object candidate) {
        return maybeToValidVector(VectorCandidate.maybeFrom(candidate));
    }

    public abstract float[] maybeToValidVector(VectorCandidate candidate);

    public float compare(float[] vector1, float[] vector2) {
        return toLucene().compare(vector1, vector2);
    }

    final org.apache.lucene.index.VectorSimilarityFunction toLucene() {
        return switch (this) {
            case EUCLIDEAN -> org.apache.lucene.index.VectorSimilarityFunction.EUCLIDEAN;
            case COSINE -> org.apache.lucene.index.VectorSimilarityFunction.COSINE;
        };
    }
}

File: community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/commands/expressions/VectorSimilarityFunction.scala
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.neo4j.cypher.internal.runtime.interpreted.commands.expressions

import org.neo4j.cypher.internal.runtime.ReadableRow
import org.neo4j.cypher.internal.runtime.interpreted.commands.AstNode
import org.neo4j.cypher.internal.runtime.interpreted.pipes.QueryState
import org.neo4j.cypher.operations.CypherFunctions.vectorSimilarity
import org.neo4j.kernel.api.impl.schema.vector.VectorSimilarity
import org.neo4j.values.AnyValue

case class VectorSimilarityFunction(similarity: VectorSimilarity, v1: Expression, v2: Expression)
    extends Expression {

  override def apply(row: ReadableRow, state: QueryState): AnyValue =
    vectorSimilarity(similarity, v1(row, state), v2(row, state))

  override def rewrite(f: Expression => Expression): Expression =
    f(VectorSimilarityFunction(similarity, v1.rewrite(f), v2.rewrite(f)))

  override def arguments: collection.Seq[Expression] = v1.arguments ++ v2.arguments

  override def children: collection.Seq[AstNode[_]] = Seq(v1, v2)

  override def toString: String = s"VectorSimilarity${similarity.name.capitalize}(${v1}, ${v2})"
}
